### ISSUE TITLE
Virtual Button index off by one for values greater than 13

### ISSUE THREAD

**Issue by fauxmight** [Vantskruv/wejoy#4]
First, I absolutely love this codebase; THANK YOU! I spend more time with wejoy than I do using the virtual devices that wejoy makes just because the power and control wejoy offers is amazing. : )

Since a recent kernel update (I am grasping at straws and guessing that this is the issue), virtual devices with more than 14 buttons (0 to 13 are ok) exhibit an odd behaviour:

`send_button_event(x, y, z)` sends a button event for button y-1 if y > 13.

For example:

`send_button_event(0, 13, 1)`     works correctly and sends a button press for button 13, but
`send_button_event(0, 14, 1)`     ALSO sends a button press for virtual device 0 button 13, and
`send_button_event(0, 15, 1)`     sends a button press for virtual device 0 button 14

One can also see using jstest that for virtual devices created with <=14 buttons, the correct number of buttons exist for the virtual device (if `buttons = 14` in the created v_devices section, then jstest shows buttons 0 through 13), but if a virtual device is created with more than 14 buttons, jstest shows one fewer button than should exist (if `buttons = 15` in the created v_devices section, then jstest still only shows buttons 0 through 13 and if `buttons = 16` then jstest shows buttons 0 through 14).

I'm on Arch Linux using kernel 4.7.6-1-ARCH. I've tested with more than one device and found the same behaviour. I can work around this by just never sending button 14 presses, but if I am not the only one being affected, I cannot determine what causes this sometimes-off-by-one behaviour. In the past, I never ran into this issue.

Please find attached an example lua config for just the throttle portion of the X-56 Rhino that exhibits this behaviour for me.

[test-lua-config.zip](https://github.com/Vantskruv/wejoy/files/517828/test-lua-config.zip)

**Comment by Vantskruv** [comment]
This seems to be hard nut to crack. I'm currently on kernel 4.6.0-1-AMD64 (running Debian Testing).
It works without issues for me, though I've tested this with the Warthog Throttle device.

Some investigation is needed to see what has been changed in your new kernel. Either there is a hidden bug in the wejoy, or a bug in your new kernel, or there is back-compability issues (deprecation?).

I'm glad you are having use of this tool, though I've not looked into code for a while. But I'll try to see what can be done, though it may be hard to get fix as I'm not running kernel-version of yours.

**Comment by fauxmight** [comment]
Thank you for the reply. I'm having trouble reproducing this now (yesterday) with anything other than the (separate USB connections) throttle and stick of the X-56. I'm at work just now. When I'm home, I will try this on Debian kernels and give you as much info as I can. If I can't reproduce this behaviour on any other device though, I'll close the issue.

**Comment by fauxmight** [comment]
OK. I got the chance to do some more testing. Using a Debian live CD (kernel 3.16.0-4-amd64), and a Saitek X-52, I still get the same result. I'm attaching the lua config here, but this is with both the X-52 and the X-56 in Arch and in Debian.

You can see in the lua config, I've just mapped physical buttons 0 through 18 to virtual buttons 0 through 18 (19 buttons). jstest reports 18 (not 19) buttons on the virtual device and pressing physical button 13 OR 14 generates a virtual button press 13. Every virtual button mapped that has an index higher than 13 generates a jstest button index press one lower than it should (mapped virtual 14 generates button press 13; mapped virtual 15 generates button press 14 and so on).

What hardware (joystick/HOTAS) are you testing on?
[test-x52.zip](https://github.com/Vantskruv/wejoy/files/531288/test-x52.zip)

**Comment by fauxmight** [comment]
Here, not really part of the issue, but there's nowhere else to just say "thanks, this project is awesome." I bolted down a couple of old sticks I had laying around, added them to my CH FighterStick/Throttle combo and using wejoy I now have a free pedal-rudder axis. : )

![logitech-pedals](https://cloud.githubusercontent.com/assets/942748/19629608/7fc264f0-9946-11e6-8dae-73c2698a8980.jpg)

**Comment by Vantskruv** [comment]
Sorry I've not replied in a while. My daily job sometimes do take my energy out.

I'll confess that have not looked into the code. But looking through it, and checking the other sources this project relies on, there may be either of two causes your bug appears.

1. It may be a button identifier from the uinput header files that has been deprecated (currently I do not have any time checking this). 
I'm not sure if you are experience with some programming, but looking into the buttons_ref.h header file, try remove the BTN_HAPPY, or the BTN_DEAD in the BUTTONS[] index, and then decrease the BUTTON_SIZE with the amount of buttons identifers you removed.

I.e. change this code:
`#define BUTTONS_SIZE 54`
to
`#define BUTTONS_SIZE 53`
and then remove either BTN_DEAD or BTN_HAPPY from the BUTTONS[] array from below.

2.
Another interesting thing is that developer of libsuinput have update the code, and maybe this fixes the bug happening for you.

I'll we research this later on when I've some more freetime.

PS.
I like your cheapmans solution for rudderpedals. :) I think this may help armless people to be more dangerous (sorry for very bad humour).
.DS

**Comment by fauxmight** [comment]
Hi, thanks for the feedback. I took a look at libsuinput. The two files from the project that you included, suinput.c and suinput.h have not been modified in the git project for the past three years. I will take a look at modifying the BTN_HAPPY and BTN_DEAD entries in the BUTTONS[] index as soon as I can.

I think the dangerous, armless people humour is great.

**Comment by fauxmight** [comment]
You pointed me in the right direction!

BTN_TRIGGER_HAPPY and BTN_TRIGGER_HAPPY1 both generate the same input event code. By removing BTN_TRIGGER_HAPPY, the duplicate issue is resolved.

I created a pull request with buttons_ref.h corrected.

Unless you see any problems with this, please feel free to close this issue after the pull is committed.

**Comment by fauxmight** [comment]
The second commit in the pull request is minor and unrelated. This just shuts up unsigned int to signed int comparison warnings during compile.

**Comment by Vantskruv** [comment]
Everything seems to be alright. :) But it was strange that I did not have any problems testing your profile (though converted to Warthog Throttle). Probably lazy testing and therefore did not see the error show up.

I'm glad this worked out. I merged the changes to this project. Feel free just to come up with some ideas or more fixes.

I did create Wejoy mainly for just being able to radioswitch button on my joystick to a key for talking to ATC in X-Plane, but currently I'm running Windows just because TrackIR does not support Linux, so therefore I've not put any energy for this tool.

Closing this issue!
